### PR TITLE
First cut at IDENTITY-izing Foursight.

### PR DESCRIPTION
**Pull request by dmichaels-harvard**
This branch also contains some changes/fixes from Kent (from: PR-27, branch: kmp_valid_env), to app_utils.py, check_utils.py, and environment.py.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -7,6 +7,13 @@ Change Log
 ----------
 
 
+0.7.4
```
Comment (willronchetti) [inline CHANGELOG.rst:35]: Should definitely be major version IMO

File: foursight_core/app_utils.py [hunk 2]
```diff
@@ -59,7 +65,47 @@ class AppUtilsCore:
...
+        apply_identity(identity_kind=GLOBAL_APPLICATION_CONFIGURATION, rename_keys=IDENTITY_KEY_MAP)
+
+        # Get the RDS secrets name; could pass this in via envrionment variable
+        # in Foursight CloudFormation template, like IDENTITY.
+        # We look here only for the RDS_NAME value; should probably move to GAC.
+        # And if were to get it from RDS secrets then should probably pass that secrets name in.
+        # TODO
+        rds_secrets_name = identity_name.replace("ApplicationConfiguration", "RDSSecret")
+        rds_secrets = SecretsTable(rds_secrets_name)
+        os.environ["RDS_NAME"] = rds_secrets.get("dbInstanceIdentifier")
```
Comment (willronchetti) [inline foursight_core/app_utils.py]: Probably easier to just add RDS_NAME to the GAC

File: foursight_core/app_utils.py [hunk 3]
```diff
@@ -59,7 +65,47 @@ class AppUtilsCore:
...
 
+    # dmichaels/2022-07-20: New to apply the IDENTITY values globally to os.environ (C4-826).
+    def apply_identity_globally(self):
+
+        # This maps key names in the global application configuration (GAC) to names used here.
+        IDENTITY_KEY_MAP = {
+            "ENCODED_AUTH0_CLIENT": "CLIENT_ID",
+            "ENCODED_AUTH0_SECRET": "CLIENT_SECRET",
+            "ENCODED_S3_ENCRYPT_KEY_ID": "S3_ENCRYPT_KEY_ID",
+            "ENCODED_ES_SERVER": "ES_HOST",
```
Comment (willronchetti) [inline foursight_core/app_utils.py]: Not a thing for now, but we might consider adding `FOURSIGHT_ES_SERVER` so these values can be different (they are right now as it is set via argument in 4dn-cloud-infra).

**Review by willronchetti — COMMENTED** [review]
Some initial comments - definitely worth a try with a deploy

File: CHANGELOG.rst (hunk 1 above)
Comment (dmichaels-harvard) [reply CHANGELOG.rst:35]: Major as in '1.0.0'? (Or just '0.8.0')?

File: foursight_core/app_utils.py (hunk 2 above)
Comment (dmichaels-harvard) [reply foursight_core/app_utils.py]: Made 4dn-cloud-infra changes to put RDS_NAME in GAC. Left this here (slightly modified) for now as a fallback if not found.

File: CHANGELOG.rst (hunk 1 above)
Comment (netsettler) [reply CHANGELOG.rst:35]: @willronchetti, not sure how you feel about this but I might say 1.0 just so we start using proper semantic versioning?

**Review by netsettler — COMMENTED** [review]
The changes look OK to me. Does it run? :) If tests pass and the system appears to work normally, I'm good with this, but you should probably get Will to approve, too.

File: foursight_core/deploy.py [hunk 4]
```diff
@@ -50,58 +50,76 @@ def get_config_filepath(cls):
...
-                sys.exit()
+            # es_host = None  # not previously passed to config
+            # env_name = None
+            # s3_key_id = None  # not supported in legacy
+            # if not (s3_enc_secret and client_id and client_secret and dev_secret):
+            #     print(''.join(['ERROR. You are missing one more more environment ',
+            #                    'variables needed to deploy Foursight.\n',
+            #                    'Need: S3_ENCRYPT_KEY, CLIENT_ID, CLIENT_SECRET, DEV_SECRET.'])
+            #           )
+            #     sys.exit()
```
Comment (willronchetti) [inline foursight_core/deploy.py:97]: I think we can remove this commented out code now (including DEV_SECRET).

File: foursight_core/deploy.py [hunk 5]
```diff
@@ -50,58 +50,76 @@ def get_config_filepath(cls):
...
-            if s3_key_id:
-                curr_stage_environ['S3_ENCRYPT_KEY_ID'] = s3_key_id
+            # dmichaels/2022-07-22/C4-826:
+            # Removed these from the Foursight CloudFormation template; get from GAC/etc at runtime.
+            # if env_name:
+            #     curr_stage_environ['ENV_NAME'] = env_name
+            # if es_host:
+            #     curr_stage_environ['ES_HOST'] = es_host
+            # if s3_key_id:
+            #     curr_stage_environ['S3_ENCRYPT_KEY_ID'] = s3_key_id
```
Comment (willronchetti) [inline foursight_core/deploy.py:122]: Should delete as well.

File: foursight_core/identity.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        rds_secrets_name = identity_name.replace("ApplicationConfiguration", "RDSSecret")
+        rds_secrets = SecretsTable(rds_secrets_name)
+        os.environ["RDS_NAME"] = rds_secrets.get("dbInstanceIdentifier")
+
+    # Get the CHECK_RUNNER lambda function name; using the stack_name as a prefix; for example,
+    # this lambda name looks like: c4-foursight-cgap-supertest-stack-CheckRunner-pKsOvziDT7QI
+    # where c4-foursight-cgap-supertest-stack is the stack_name.
+    check_runner_lambda_function_name_pattern = stack_name + "-CheckRunner-.*"
+    check_runner_lambda_function_names = (
+        AbstractOrchestrationManager.find_lambda_function_names(check_runner_lambda_function_name_pattern))
```
Comment (willronchetti) [inline foursight_core/identity.py:75]: Bit of an odd spacing pattern (a kwarg may help too), I would say:
```
check_runner_lambda_function_names = AbstractOrchestrationManager.find_lambda_function_names(
    check_runner_lambda_function_name_pattern
)
```

File: foursight_core/s3_connection.py [hunk 7]
```diff
@@ -56,6 +56,9 @@ def get_object(self, key):
                 return json.loads(body)
             except json.JSONDecodeError:
                 return body
+        except self.client.exceptions.NoSuchKey as e_no_such_key:
+            logger.error(f"No such key found for S3 bucket ({self.bucket}): {key}")
+            logger.error(e_no_such_key)
```
Comment (willronchetti) [inline foursight_core/s3_connection.py:61]: I would log this as one statement

**Review by willronchetti — APPROVED** [review]
Approving considering this deploys and looks good to me with some minor comments

**Pull request merged**